proposals/lip-0053.md

```diff
@@ -133,28 +144,247 @@ The following constants are used throughout the document:
+        certificateBytes = trs.params.certificate
+        if len(certificateBytes) == 0:
+            raise Exception("The first mainchain CCU must contain a non-empty certificate.")
+        
+        certificate = decode(certificateSchema, certificateBytes)
+        timestamp = timestamp of the block including CCU
+        if len(trs.params.inboxUpdate) > 0 and timestamp - certificate.timestamp > LIVENESS_LIMIT / 2:
+            raise Exception(f"First certificate cannot be older than {LIVENESS_LIMIT / 2} seconds.")
+
+    elif not isLive(sidechainID):
+        raise Exception("Sidechain is not live.")
+``` 
+
+#### verifyCertificate
+
+```python
+def verifyCertificate(trs: Transaction) -> None:
+    certificate = decode(certificateSchema, trs.params.certificate)
+    partnerchainID = trs.params.sendingChainID
+    partnerchainAccount = chainAccount(partnerchainID)
+    # Certificate height must be greater than last certificate height.
+    if certificate.height <= partnerchainAccount.lastCertificate.height:
+        raise Exception("Certificate height is not greater than last certificate height.")
+    # Certificate timestamp must be smaller than timestamp of the block including the CCU.
+    timestamp = timestamp of the block including CCU
+    if certificate.timestamp >= timestamp:
+        raise Exception("Certificate timestamp is not smaller than timestamp of the block including the CCU.")
+```
+
+#### verifyCertificateSignature
+
+```python
+def verifyCertificateSignature(trs: Transaction) -> None:
+    # Certificate signature must be a valid aggregate signature for the sdechain validators.
+    # verifyAggregateCertificateSignature is defined in LIP 0061.
+    certificate = decode(certificateSchema, trs.params.certificate)
+    partnerchainID = trs.params.sendingChainID
+    partnerchainAccount = chainAccount(partnerchainID)
+    validatorKeys = [validator.blsKey for validator in validators(partnerchainID).activeValidators]
+    validatorWeights = [validator.bftWeight for validator in validators(partnerchainID).activeValidators]
+    verifyAggregateCertificateSignature(
+        validatorKeys, 
+        validatorWeights, 
+        validators(partnerchainID).certificateThreshold, 
+        partnerchainAccount.networkID, 
+        certificate
+    )
+```
+
+#### verifyValidatorsUpdate
 
-The function `uint32be(x)` returns the big endian uint32 serialization of an integer `x`, with `0 <= x < 2^32`. This serialization is always 4 bytes long.
+```python
+def verifyValidatorsUpdate(trs: Transaction) -> None:
+    if len(trs.params.certificate) == 0:
+        raise Exception("Certificate must be non-empty if validators or certicate threshold have been updated.")
+    
+    partnerchainID = trs.params.sendingChainID
+    validatorKeys = [validator["blsKey"] for validator in validators(partnerchainID).activeValidators]
+    # All validator keys must be valid BLS keys.
+    for key in validatorKeys:
```

Comment:
Here I am checking the validator keys from the state store. Don't we want to check those in `trs.params.activeValidatorsUpdate` instead?

Reply:
yes, thanks!

Reply:
Looks good now.